sql/fulltext/fulltext.go

```diff
@@ -0,0 +1,488 @@
+func (editor FulltextEditor) StatementBegin(ctx *sql.Context) {
+	editor.Config.StatementBegin(ctx)
+	for _, editorIndex := range editor.Indexes {
+		editorIndex.WordToPos.StatementBegin(ctx)
+		editorIndex.Count.StatementBegin(ctx)
+	}
+}
+
+// DiscardChanges implements the interface TableEditor.
+func (editor FulltextEditor) DiscardChanges(ctx *sql.Context, errorEncountered error) error {
+	err := editor.Config.DiscardChanges(ctx, errorEncountered)
+	for _, editorIndex := range editor.Indexes {
+		if nErr := editorIndex.WordToPos.DiscardChanges(ctx, errorEncountered); err == nil {
+			err = nErr
+		}
+		if nErr := editorIndex.Count.DiscardChanges(ctx, errorEncountered); err == nil {
+			err = nErr
+		}
+	}
+	return err
+}
+
+// StatementComplete implements the interface TableEditor.
+func (editor FulltextEditor) StatementComplete(ctx *sql.Context) error {
+	err := editor.Config.StatementComplete(ctx)
+	for _, editorIndex := range editor.Indexes {
+		if nErr := editorIndex.WordToPos.StatementComplete(ctx); err == nil {
+			err = nErr
+		}
+		if nErr := editorIndex.Count.StatementComplete(ctx); err == nil {
+			err = nErr
+		}
+	}
+	return err
+}
+
+// Insert implements the interface TableEditor.
+func (editor FulltextEditor) Insert(ctx *sql.Context, row sql.Row) error {
+	ftsDocId := row[editor.DocIdIndex].(uint64)
+	for _, index := range editor.Indexes {
+		// A multi-column index treats a row as the concatenation of all columns, so we track our current position
+		position := uint64(0)
+		for _, sourceCol := range index.SourceCols {
+			colStr := row[sourceCol].(string) //TODO: verify that this will always be a string and never a []byte
+			words := strings.Split(colStr, " ")
+			for _, word := range words {
+				if err := index.WordToPos.Insert(ctx, sql.Row{word, ftsDocId, position}); err != nil {
+					return err
+				}
+				// We always add 1 to account for the space between words. This also works for multi-column indexes.
+				position += uint64(len(word)) + 1
+				//TODO: write to the count table
+				//TODO: write to the config table
+			}
+		}
+	}
+	return nil
+}
+
+// Update implements the interface TableEditor.
+func (editor FulltextEditor) Update(ctx *sql.Context, old sql.Row, new sql.Row) error {
+	// I'm sure a bespoke UPDATE routine would be more efficient, but this will work for now
+	if err := editor.Delete(ctx, old); err != nil {
+		return err
+	}
+	return editor.Insert(ctx, new)
+}
+
+// Delete implements the interface TableEditor.
+func (editor FulltextEditor) Delete(ctx *sql.Context, row sql.Row) error {
+	ftsDocId := row[editor.DocIdIndex].(uint64)
+	for _, index := range editor.Indexes {
+		// A multi-column index treats a row as the concatenation of all columns, so we track our current position
+		position := uint64(0)
+		for _, sourceCol := range index.SourceCols {
+			colStr := row[sourceCol].(string) //TODO: verify that this will always be a string and never a []byte
+			words := strings.Split(colStr, " ")
```

Comment:
Since we need to do this in 2 (eventually 3) places, and it absolutely needs to be the same in both places...factor out the logic that turns the column value into a series of index edits and use it in both places and independently test it, etc.